### PR TITLE
Re-architect Iterator class.

### PR DESCRIPTION
**Pull request by dhermes**
This PR should **definitely** reviewed commit-by-commit.

Primary changes are:

- Have an iterable `Page` base-class for iterating over a single page of a result set
- Move most custom iterator logic back up into the base page and iterator classes

### CODE REVIEW

File: core/google/cloud/iterator.py [hunk 1]
```diff
@@ -38,25 +42,114 @@ def get_items_from_response(self, response):
...
+        :rtype: int
+        :returns: The number of items remaining this page.
+        """
+        return self._remaining
+
+    def __iter__(self):
+        """The :class:`Page` is an iterator."""
+        return self
+
+    def _item_to_value(self, item):
```
Comment (jonparrott) [inline core/google/cloud/iterator.py]: You could use `abc.abstractmethod` if you wanted to - just a suggestion, no action needed.

File: core/google/cloud/iterator.py [hunk 2]
```diff
@@ -38,25 +42,114 @@ def get_items_from_response(self, response):
...
+    def num_items(self):
+        """Total items in the page.
+
+        :rtype: int
+        :returns: The number of items in this page of items.
+        """
+        return self._num_items
+
+    @property
+    def remaining(self):
```
Comment (jonparrott) [inline core/google/cloud/iterator.py]: At first read I thought this was the remaining *pages*, not *items*. Maybe name this `remaining_items` to match `num_items`?

**Comment by dhermes** [comment]
> My one concern is how a service like datastore will work.

Datastore has it's own specialized `Iterator`. For the time being, the `google.cloud.iterator.Iterator` class is for JSON-based APIs that define a `nextPageToken` in response pages.

Though a "next step" would be to make the iterator support gRPC / GAPIC generated classes, so we may be able to tackle datastore there. (Just not in this PR.)

File: core/google/cloud/iterator.py (hunk 1 above)
Comment (dhermes) [inline core/google/cloud/iterator.py]: I am not "comfortable" with `abc` and don't feel like it provides anything that `raise NotImplemenetedError` doesn't

**Comment by daspecster** [comment]
Sorry I haven't responded on this @dhermes. I don't have much feedback though.

You actually just answered the question I did have.
>Though a "next step" would be to make the iterator support gRPC / GAPIC generated classes, so we may be able to tackle datastore there. (Just not in this PR.)

**Comment by jonparrott** [comment]
LGTM

**Pull request merged**